typing/ikinds/ikind.ml

```diff
@@ -0,0 +1,113 @@
+(*-------------------------------------------------------------------------*)
+(* Types shared by ikind algorithms                                        *)
+(*-------------------------------------------------------------------------*)
+
+module RigidName = struct
+  (* Rigid names are the variables that may occur in ikind formulas. *)
+
+  type t =
+    | Atom of
+        { constr : constr;
+          (* arg_index = 0 refers to the base contribution,
+             subsequent refer to the coefficients of the i-th argument. *)
+          arg_index : int
+        }
+    (* Param only occurs in the formula for a type constructor,
+       refers to the i-th parameter. *)
+    | Param of int
+    (* An unknown quantity with a given id. Used to model not-best in ikinds.
+       This is used when we couldn't compute a precise ikind, e.g. for a
+       polymorphic variant with conjunctive type -- `Constr of (a & b & ...) *)
+    | Unknown of int
+
+  and constr = Path.t
+
+  (* Ordering on rigid names used in the LDD to order the nodes. *)
+  let compare a b =
+    if a == b
+    then 0
+    else
+      match a, b with
+      | Atom a1, Atom a2 ->
+        let h = Path.compare a1.constr a2.constr in
+        if h <> 0 then h else Int.compare a1.arg_index a2.arg_index
```

Comment:
```suggestion
        if h != 0 then h else Int.compare a1.arg_index a2.arg_index
```
(These are semantically the same in this context, but I think != expresses the intent more clearly)